### PR TITLE
Add GCS support for UPI installs

### PR DESCRIPTION
**Pull request by coreydaley**
https://jira.coreos.com/browse/DEVEXP-197

### CODE REVIEW

**Comment by coreydaley** [comment]
@dmage @adambkaplan bump

File: pkg/storage/gcs/gcs.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+	if cr.Status.Storage.GCS == nil {
+		cr.Status.Storage.GCS = &imageregistryv1.ImageRegistryConfigStorageGCS{}
+	}
+	if len(d.Config.Bucket) == 0 {
+		d.Config.Bucket = cfg.Storage.GCS.Bucket
+	}
+
+	cr.Status.Storage.GCS = d.Config.DeepCopy()
+
+	return nil
```
Comment (adambkaplan) [inline pkg/storage/gcs/gcs.go]: Per https://github.com/openshift/cluster-image-registry-operator/pull/279#discussion_r285176552 - return `nil` in this function for now. @dmage is going to do a refactor later.

File: pkg/clusterconfig/clusterconfig.go [hunk 2]
```diff
@@ -152,7 +152,20 @@ func GetAWSConfig(listers *regopclient.Listers) (*Config, error) {
...
-func GetGCSConfig() (*Config, error) {
+func GetGCSConfig(listers *regopclient.Listers) (*Config, error) {
 	cfg := &Config{}
+
+	// Look for a user defined secret to get the GCS credentials from
+	sec, err := listers.Secrets.Get(imageregistryv1.ImageRegistryPrivateConfigurationUser)
+	if err != nil {
+		return nil, err
+	} else {
+		if v, ok := sec.Data["STORAGE_GCS_KEYFILE"]; ok {
```
Comment (adambkaplan) [inline pkg/clusterconfig/clusterconfig.go]: Add a comment here that GCS credentials are stored in a keyfile provided by Google. 

See also [1] and [2].

[1] https://cloud.google.com/storage/docs/authentication#service_accounts
[2] https://cloud.google.com/docs/authentication/production

File: pkg/storage/gcs/gcs.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
+package gcs
+
+import (
+	"reflect"
+
+	coreapi "k8s.io/api/core/v1"
```
Comment (adambkaplan) [inline pkg/storage/gcs/gcs.go]: /s/coreapi/corev1/

File: pkg/storage/gcs/gcs.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+		coreapi.EnvVar{Name: "REGISTRY_STORAGE_GCS_KEYFILE", Value: "/gcs/keyfile"},
+	)
+	return
+}
+
+func (d *driver) Volumes() ([]coreapi.Volume, []coreapi.VolumeMount, error) {
+	vol := coreapi.Volume{
+		Name: "registry-storage-keyfile",
+		VolumeSource: coreapi.VolumeSource{
+			Projected: &coreapi.ProjectedVolumeSource{
```
Comment (adambkaplan) [inline pkg/storage/gcs/gcs.go]: side note - `ProjectedVolumeSource` is new to me (and I like it!)

Looks like it's GA in kube 1.13, so no concerns here w.r.t upgrades.

File: pkg/clusterconfig/clusterconfig.go (hunk 2 above)
Comment (coreydaley) [reply pkg/clusterconfig/clusterconfig.go]: updated in next push

File: pkg/storage/gcs/gcs.go (hunk 3 above)
Comment (coreydaley) [reply pkg/storage/gcs/gcs.go]: updated in next push

File: pkg/storage/gcs/gcs.go [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+	if !reflect.DeepEqual(cr.Status.Storage.GCS, cr.Spec.Storage.GCS) {
+		util.UpdateCondition(cr, imageregistryv1.StorageExists, operatorapi.ConditionUnknown, "GCS Configuration Changed", "GCS storage is in an unknown state")
+		return true
+	}
+
+	return false
+}
+
+func (d *driver) CreateStorage(cr *imageregistryv1.Config) error {
+	if !reflect.DeepEqual(cr.Status.Storage.S3, d.Config) {
```
[the remark this answers is not in the document]
Comment (coreydaley) [reply pkg/storage/gcs/gcs.go]: updated in next push

File: pkg/storage/gcs/gcs.go (hunk 1 above)
Comment (coreydaley) [reply pkg/storage/gcs/gcs.go]: Updated in next push (hope it passes testing)

**Comment by coreydaley** [comment]
@adambkaplan ptal, updated per your comments

**Comment by coreydaley** [comment]
/retest

504 error level events were detected during this test run:

**Comment by coreydaley** [comment]
Rebased to fix merge conflicts

File: pkg/storage/gcs/gcs.go [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+		corev1.EnvVar{Name: "REGISTRY_STORAGE_GCS_KEYFILE", Value: "/gcs/keyfile"},
+	)
+	return
+}
+
+func (d *driver) Volumes() ([]corev1.Volume, []corev1.VolumeMount, error) {
+	vol := corev1.Volume{
+		Name: "registry-storage-keyfile",
+		VolumeSource: corev1.VolumeSource{
+			Projected: &corev1.ProjectedVolumeSource{
```
Comment (dmage) [inline pkg/storage/gcs/gcs.go]:
```
Secret: &corev1.SecretVolumeSource{
```

There is no need to use a projected volume source.

Comment (coreydaley) [reply pkg/storage/gcs/gcs.go]: Updated in current push

**Pull request merged**